File: 3-hard/01944-number-of-visible-people-in-a-queue/solution.cpp
class Solution {
public:
    vector<int> canSeePersonsCount(vector<int>& heights) {

        vector<int> res(heights.size(), 0);
        stack<int> stack{};

        for (int i{(int)heights.size() - 1}; i >= 0; --i) {
            while (!stack.empty()) {

                res[i] += 1;

                int val = stack.top();
                if (val <= heights[i]) {
                    stack.pop();
                } else {
                    break;
                }

            }

            stack.push(heights[i]);
        }

        return res;
    }
};

File: 3-hard/01944-number-of-visible-people-in-a-queue/solution.rs
impl Solution {
    pub fn can_see_persons_count(heights: Vec<i32>) -> Vec<i32> {

        let mut res: Vec<i32> = vec![0; heights.len()];
        let mut stack: Vec::<i32> = Vec::new();

        for i in (0..heights.len()).rev() {
            while let Some(val) = stack.last() {
                res[i] += 1;
                if *val <= heights[i] {
                    stack.pop();
                } else {
                    break;
                }
            }
            stack.push(heights[i]);
        }

        res
    }
}
